src/libaktualizr/primary/aktualizr_test.cc

```diff
@@ -492,6 +500,358 @@ TEST(Aktualizr, FullWithUpdatesNeedReboot) {
+        storage->loadInstalledVersions(primary_ecu_id, &installed_versions, &current_version, &pending_version));
+
+    // for some reason there is no any installed version at initial Aktualizr boot/run
+    // IMHO it should return currently installed version
+    ASSERT_TRUE(installed_versions.empty());
+    ASSERT_EQ(pending_version, SIZE_MAX);
+    ASSERT_EQ(current_version, SIZE_MAX);
+
+    auto aktualizr_cycle_thread = aktualizr.RunForever();
+    auto aktualizr_cycle_thread_status = aktualizr_cycle_thread.wait_for(std::chrono::seconds(10));
+
+    ASSERT_EQ(aktualizr_cycle_thread_status, std::future_status::ready);
+    ASSERT_TRUE(aktualizr.uptane_client_->bootloader->rebootDetected());
+    ASSERT_TRUE(event_hdlr.checkRecievedEvents(expected_event_order));
+    ASSERT_TRUE(aktualizr.uptane_client_->hasPendingUpdates());
+    ASSERT_TRUE(http_server_mock->checkRecievedReports(expected_report_order));
+    // Aktualizr reports to a server that installation was successfull for the secondary
+    // checkRecievedReports() verifies whether EcuInstallationApplied was reported for the primary
+    ASSERT_FALSE(http_server_mock->wasInstallSuccessful(primary_ecu_id));
+    ASSERT_TRUE(http_server_mock->wasInstallSuccessful(secondary_ecu_id));
+
+    data::InstallationResult dev_installation_res;
+    std::string report;
+    std::string correlation_id;
+    ASSERT_TRUE(storage->loadDeviceInstallationResult(&dev_installation_res, &report, &correlation_id));
+    EXPECT_EQ(dev_installation_res.result_code.num_code, data::ResultCode::Numeric::kNeedCompletion);
+
+    std::vector<std::pair<Uptane::EcuSerial, data::InstallationResult>> ecu_installation_res;
+    EXPECT_TRUE(storage->loadEcuInstallationResults(&ecu_installation_res));
+    EXPECT_EQ(ecu_installation_res.size(), 2);
+
+    for (const auto& ecu_install_res : ecu_installation_res) {
+      auto ecu_id = ecu_install_res.first.ToString();
+
+      if (ecu_id == primary_ecu_id) {
+        EXPECT_EQ(ecu_install_res.second.result_code.num_code, data::ResultCode::Numeric::kNeedCompletion);
+        EXPECT_EQ(ecu_install_res.second.success, false);
+        EXPECT_EQ(ecu_install_res.second.description, "Application successful, need reboot");
+
+      } else if (ecu_id == secondary_ecu_id) {
+        EXPECT_EQ(ecu_install_res.second.result_code.num_code, data::ResultCode::Numeric::kOk);
+        EXPECT_EQ(ecu_install_res.second.success, true);
+        EXPECT_EQ(ecu_install_res.second.description, "");
+
+      } else {
+        FAIL() << "Unexpected ECU serial: " << ecu_install_res.first;
+      }
+    }
+
+    pending_version = SIZE_MAX;
+    current_version = SIZE_MAX;
+
+    ASSERT_TRUE(
+        storage->loadInstalledVersions(primary_ecu_id, &installed_versions, &current_version, &pending_version));
+    ASSERT_EQ(installed_versions.size(), 1);
+    ASSERT_TRUE(installed_versions[0].IsValid());
+    // if pending_version equals 0 then it means that this is a pending version
```

Comment:
So 0 really means "yes" for these? That is weird, or at least, not what I would've expected. Not that it's your fault or anything.

Reply:
Yes, 0 means "yes" in this case since loadInstalledVersions() returns index of 'current_version' and 'pending_version' in the returned array 'installed_versions'. If there is no such version (pending and/or current) then SIZE_MAX is returned as its index value.

Reply:
Now I understand. Thanks!